### PR TITLE
[SDK-2576] Fix blue dot delay

### PR DESCRIPTION
**Pull request by ahmetpointr**
<changelog>Fix blue dot being delayed when map tracking mode is one of the follow modes.</changelog>

### CODE REVIEW

File: platform/darwin/src/MGLLocationManager.h [hunk 1]
```diff
@@ -165,6 +165,20 @@ NS_ASSUME_NONNULL_BEGIN
...
+ @param manager The location manager reporting the update.
+ @param locations An array of `CLLocation` objects in chronological order,
+ with the last object representing the most recent location. This array
+ contains multiple `CLLocation` objects when `MGLMapView` uses  its
+ default location manager.
+ @param heading The heading update.
+ */
+- (void)locationManager:(id<MGLLocationManager>)manager
+     didUpdateLocations:(NSArray<CLLocation *> *)locations
+                heading:(nullable CLHeading *)heading;
```
Comment (manuyavuz-pointr) [inline platform/darwin/src/MGLLocationManager.h:177]: This method's API is a little bit inconsistent as if we provide a location array, it's expected to provide a heading array too.

I think we should have a method with a single `CLLocation` parameter. Mapbox internally ignores all but the last `CLLocation` object in the array anyways.

File: platform/ios/src/MGLMapView.mm [hunk 2]
```diff
@@ -1131,7 +1131,7 @@ - (void)setContentInset:(UIEdgeInsets)contentInset animated:(BOOL)animated compl
     else
     {
         _contentInset = contentInset;
-        [self didUpdateLocationWithUserTrackingAnimated:animated completionHandler:completion];
+        [self didUpdateLocationWithUserTrackingHeading:nil animated:animated completionHandler:completion];
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm]: let's keep having `didUpdateLocationWithUserTrackingAnimated: completionHandler:` a convenience method which internally calls newly added `didUpdateLocationWithUserTrackingHeading:nil animated: completionHandler:` so that we don't change every place where heading is not available.

File: platform/ios/src/MGLMapView.mm [hunk 3]
```diff
@@ -5657,7 +5657,7 @@ - (void)setUserTrackingMode:(MGLUserTrackingMode)mode animated:(BOOL)animated co
     CLLocation *location;
     if (_userTrackingMode != MGLUserTrackingModeNone && (location = self.userLocation.location) && self.userLocationAnnotationView)
     {
-        [self locationManager:self.locationManager didUpdateLocations:@[location] animated:animated completionHandler:completion];
+        [self locationManager:self.locationManager didUpdateLocations:@[location] heading:nil animated:animated completionHandler:completion];
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm]: same here, let's keep the original method as convenience, which internally calls the method with newly added `heading:` parameter.

File: platform/ios/src/MGLMapView.mm [hunk 4]
```diff
@@ -5754,10 +5754,15 @@ - (void)validateUserHeadingUpdating
 
 - (void)locationManager:(id<MGLLocationManager>)manager didUpdateLocations:(NSArray *)locations
 {
-    [self locationManager:manager didUpdateLocations:locations animated:YES completionHandler:nil];
+    [self locationManager:manager didUpdateLocations:locations heading:nil];
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm]: same here, keep original method as convenience.

File: platform/ios/src/MGLMapView.mm [hunk 5]
```diff
@@ -5810,6 +5815,15 @@ - (void)didUpdateLocationWithUserTrackingAnimated:(BOOL)animated completionHandl
...
 
+    CLLocationDirection direction = self.directionByFollowingWithCourse;
+    if (heading != nil && heading.headingAccuracy > 0) {
+        self.userLocation.heading = heading;
+        if (self.userTrackingMode == MGLUserTrackingModeFollowWithHeading)
+        {
+            direction = (heading.trueHeading >= 0 ? heading.trueHeading : heading.magneticHeading);
+        }
+    }
+
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm]: why do we have this logic in this `didUpdateLocationWithUserTrackingAnimated:` method? it seems we only need this in `- (void)locationManager:(__unused id<MGLLocationManager>)manager didUpdateLocations:(NSArray *)locations heading:(nullable CLHeading *)heading animated:(BOOL)animated completionHandler:(nullable void (^)(void))completion` method, so we do this logic directly in that method, similar to what is currently available in `- (void)locationManager:(__unused id<MGLLocationManager>)manager didUpdateHeading:(CLHeading *)newHeading` method.

Comment (manuyavuz-pointr) [reply platform/ios/src/MGLMapView.mm]: Also, it seems like we simply ignored a few checks that exist in `locationManager:didUpdateHeading:` method. can you explain the need to ignore these checks?

File: platform/ios/src/MGLMapView.mm [hunk 6]
```diff
@@ -5830,38 +5844,38 @@ - (void)didUpdateLocationWithUserTrackingAnimated:(BOOL)animated completionHandl
         if (self.userTrackingState != MGLUserTrackingStateBegan)
         {
             // Keep both the user and the destination in view.
-            [self didUpdateLocationWithTargetAnimated:animated completionHandler:completion];
+            [self didUpdateLocationWithTargetDirection:direction animated:animated completionHandler:completion];
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm]: Rather than changing all of these methods to have the direction parameter, i think it's better to make `directionByFollowingWithCourse` a readwrite property, and use it as previous in these methods.

File: platform/ios/src/MGLMapView.mm [hunk 7]
```diff
@@ -6694,16 +6708,7 @@ - (void)updateUserLocationAnnotationViewAnimatedWithDuration:(NSTimeInterval)dur
...
-    if (self.userTrackingMode != MGLUserTrackingModeNone
-        && self.userTrackingState == MGLUserTrackingStateChanged)
-    {
-        userPoint = self.userLocationAnnotationViewCenter;
-    }
-    else
-    {
-        userPoint = MGLPointRounded([self convertCoordinate:self.userLocation.coordinate toPointToView:self]);
-    }
+    CGPoint userPoint = MGLPointRounded([self convertCoordinate:self.userLocation.coordinate toPointToView:self]);
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm:6722]: Why this change is needed? What's the difference in behavior?

File: platform/ios/src/MGLMapView.mm [hunk 8]
```diff
@@ -5889,7 +5903,7 @@ - (void)didUpdateLocationSignificantlyAnimated:(BOOL)animated completionHandler:
      completionHandler:^{
         MGLMapView *strongSelf = weakSelf;
         if (strongSelf.userTrackingState == MGLUserTrackingStateBegan ||
-            strongSelf.userTrackingState == MGLDistanceThresholdForCameraPause)
+            strongSelf.userTrackingState == MGLUserTrackingStateBeginSignificantTransition)
```
Comment (manuyavuz-pointr) [inline platform/ios/src/MGLMapView.mm:5905]: nice catch 👍 , and what a bug :D

**Review by manuyavuz-pointr — CHANGES_REQUESTED** [review]
@ahmetpointr I have a few questions, and some suggestions.

In general, we should try to keep the api the same as much as possible, to simplify any future rebase onto new versions and upstream changes.

**Comment by manuyavuz-pointr** [comment]
@ahmetpointr can you also explain here what was the exact behavior change in terms of code flow that fixed the delay issue?

I see that with providing location and heading simultaneously, we skip the headingUpdate call, which internally cancels all previous animation transactions and triggers `mbglMap.easeTo` with a fixed animation duration.

Am I reading this correctly?

File: platform/darwin/src/MGLLocationManager.h (hunk 1 above)
Comment (ahmetpointr) [reply platform/darwin/src/MGLLocationManager.h:177]: 👍 I can change this method to have a single location.

File: platform/ios/src/MGLMapView.mm (hunk 7 above)
Comment (ahmetpointr) [reply platform/ios/src/MGLMapView.mm:6722]: Actually removed lines were causing the delay issue. `self.userLocationAnnotationViewCenter` is a computed property and it computes the user location wrong during transition. 

https://github.com/mapbox/mapbox-gl-native-ios/commit/453000af3b17a8776ff0e78593dd5682c8cdab8d#diff-2b12ba2cacef0e8c6a943690a80480a938985ff758d0b69cae5913c8d3cf9273R3395

The computed property is added in this commit to add margin to center from blue dot in order to have this:
// When tracking course, it’s more important to see the road ahead, so weight the user dot down towards the bottom.

I believe commits after this commit (~ 5 years) may have been broke the experience.

File: platform/ios/src/MGLMapView.mm (hunk 2 above)
Comment (ahmetpointr) [reply platform/ios/src/MGLMapView.mm]: These methods are all called by `didUpdateLocations` method to which I added another version. If we keep them as is, we should have our implementations of these methods.

File: platform/ios/src/MGLMapView.mm (hunk 6 above)
Comment (ahmetpointr) [reply platform/ios/src/MGLMapView.mm]: `directionByFollowingWithCourse` is also a computed property.

File: platform/ios/src/MGLMapView.mm (hunk 5 above)
Comment (ahmetpointr) [reply platform/ios/src/MGLMapView.mm]: `updateUserLocationAnnotationView` `self.delegate mapView:self didUpdateUserLocation:self.userLocation` and `mbglMap.easeTo` methods are already called in `didUpdateLocations` method. So the only thing what is left was `direction`.

File: platform/ios/src/MGLMapView.mm (hunk 8 above)
Comment (ahmetpointr) [reply platform/ios/src/MGLMapView.mm:5905]: this bug is still in place in master of mapbox repo :)

**Review by manuyavuz-pointr — APPROVED** [review]
